### ISSUE TITLE
HTTP respoonse has a malformed body [BUG]

### ISSUE THREAD

**Issue by bryandmc** [Azure/azure-sdk-for-java#9982]
**Describe the bug**
Version 12.5.0 of the azure-storage-blob library throws this exception (below) upon simple usage of the SDK. It's somewhat unclear because of the stack below which of 2 possible calls to the SDK caused this. It's either after calling `blobContainerClient.getBlobContainerurl()` or `blobServiceClient.generateAccountSas(..)`. These calls all worked under the previous version we used (12.3.0). We have changed our dependency back to that in the short-term to deal with the issue. Unfortunately the code in question is closed-source so I can't provide long detailed snippets from our codebase, but it should be naively reproducible by just using that version with one of these calls. It happens every time in our situation. 

***Exception or Stack Trace***
```
com.azure.core.exception.HttpResponseException: HTTP response has a malformed body.
	at com.azure.core.implementation.serializer.HttpResponseBodyDecoder.lambda$decode$1(HttpResponseBodyDecoder.java:95) ~[factory.jar:master]
	at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:118) ~[factory.jar:master]
	at reactor.core.publisher.FluxDoFinally$DoFinallySubscriber.onNext(FluxDoFinally.java:123) ~[factory.jar:master]
	at reactor.core.publisher.FluxHandle$HandleSubscriber.onNext(FluxHandle.java:112) ~[factory.jar:master]
	at reactor.core.publisher.FluxMap$MapConditionalSubscriber.onNext(FluxMap.java:213) ~[factory.jar:master]
	at reactor.core.publisher.FluxDoFinally$DoFinallySubscriber.onNext(FluxDoFinally.java:123) ~[factory.jar:master]
	at reactor.core.publisher.FluxHandleFuseable$HandleFuseableSubscriber.onNext(FluxHandleFuseable.java:178) ~[factory.jar:master]
	at reactor.core.publisher.FluxContextStart$ContextStartSubscriber.onNext(FluxContextStart.java:103) ~[factory.jar:master]
	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1592) ~[factory.jar:master]
	at reactor.core.publisher.MonoCollectList$MonoCollectListSubscriber.onComplete(MonoCollectList.java:121) ~[factory.jar:master]
	at reactor.core.publisher.FluxPeek$PeekSubscriber.onComplete(FluxPeek.java:252) ~[factory.jar:master]
	at reactor.core.publisher.FluxMap$MapSubscriber.onComplete(FluxMap.java:136) ~[factory.jar:master]
	at reactor.netty.channel.FluxReceive.terminateReceiver(FluxReceive.java:419) ~[factory.jar:master]
	at reactor.netty.channel.FluxReceive.drainReceiver(FluxReceive.java:209) ~[factory.jar:master]
	at reactor.netty.channel.FluxReceive.onInboundComplete(FluxReceive.java:367) ~[factory.jar:master]
	at reactor.netty.channel.ChannelOperations.onInboundComplete(ChannelOperations.java:363) ~[factory.jar:master]
	at reactor.netty.channel.ChannelOperations.terminate(ChannelOperations.java:412) ~[factory.jar:master]
	at reactor.netty.http.client.HttpClientOperations.onInboundNext(HttpClientOperations.java:572) ~[factory.jar:master]
	at reactor.netty.channel.ChannelOperationsHandler.channelRead(ChannelOperationsHandler.java:90) ~[factory.jar:master]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:377) ~[factory.jar:master]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363) ~[factory.jar:master]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:355) ~[factory.jar:master]
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436) ~[factory.jar:master]
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:321) ~[factory.jar:master]
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:295) ~[factory.jar:master]
	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251) ~[factory.jar:master]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:377) ~[factory.jar:master]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363) ~[factory.jar:master]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:355) ~[factory.jar:master]
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1484) ~[factory.jar:master]
	at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1233) ~[factory.jar:master]
	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1280) ~[factory.jar:master]
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:498) ~[factory.jar:master]
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:437) ~[factory.jar:master]
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:276) ~[factory.jar:master]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:377) ~[factory.jar:master]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363) ~[factory.jar:master]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:355) ~[factory.jar:master]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410) ~[factory.jar:master]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:377) ~[factory.jar:master]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363) ~[factory.jar:master]
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919) ~[factory.jar:master]
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163) ~[factory.jar:master]
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714) ~[factory.jar:master]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650) ~[factory.jar:master]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576) ~[factory.jar:master]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493) ~[factory.jar:master]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989) ~[factory.jar:master]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[factory.jar:master]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[factory.jar:master]
	at java.lang.Thread.run(Thread.java:834) [?:?]
Caused by: com.azure.core.implementation.serializer.MalformedValueException: Unexpected first character (char code 0xEF), not valid in xml document: could be mangled UTF-8 BOM marker. Make sure that the Reader uses correct encoding or pass an InputStream instead
	at com.azure.core.util.serializer.JacksonAdapter.deserialize(JacksonAdapter.java:184) ~[factory.jar:master]
	at com.azure.core.implementation.serializer.HttpResponseBodyDecoder.deserializeBody(HttpResponseBodyDecoder.java:176) ~[factory.jar:master]
	at com.azure.core.implementation.serializer.HttpResponseBodyDecoder.lambda$decode$1(HttpResponseBodyDecoder.java:88) ~[factory.jar:master]
	... 50 more
Caused by: com.fasterxml.jackson.core.JsonParseException: Unexpected first character (char code 0xEF), not valid in xml document: could be mangled UTF-8 BOM marker. Make sure that the Reader uses correct encoding or pass an InputStream instead
	at com.fasterxml.jackson.dataformat.xml.util.StaxUtil.throwAsParseException(StaxUtil.java:37) ~[factory.jar:master]
	at com.fasterxml.jackson.dataformat.xml.XmlFactory._createParser(XmlFactory.java:584) ~[factory.jar:master]
	at com.fasterxml.jackson.dataformat.xml.XmlFactory.createParser(XmlFactory.java:464) ~[factory.jar:master]
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3205) ~[factory.jar:master]
	at com.azure.core.util.serializer.JacksonAdapter.deserialize(JacksonAdapter.java:179) ~[factory.jar:master]
	at com.azure.core.implementation.serializer.HttpResponseBodyDecoder.deserializeBody(HttpResponseBodyDecoder.java:176) ~[factory.jar:master]
	at com.azure.core.implementation.serializer.HttpResponseBodyDecoder.lambda$decode$1(HttpResponseBodyDecoder.java:88) ~[factory.jar:master]
	... 50 more
Caused by: com.ctc.wstx.exc.WstxIOException: Unexpected first character (char code 0xEF), not valid in xml document: could be mangled UTF-8 BOM marker. Make sure that the Reader uses correct encoding or pass an InputStream instead
	at com.ctc.wstx.io.ReaderBootstrapper.bootstrapInput(ReaderBootstrapper.java:175) ~[factory.jar:master]
	at com.ctc.wstx.stax.WstxInputFactory.doCreateSR(WstxInputFactory.java:573) ~[factory.jar:master]
	at com.ctc.wstx.stax.WstxInputFactory.createSR(WstxInputFactory.java:633) ~[factory.jar:master]
	at com.ctc.wstx.stax.WstxInputFactory.createSR(WstxInputFactory.java:687) ~[factory.jar:master]
	at com.ctc.wstx.stax.WstxInputFactory.createXMLStreamReader(WstxInputFactory.java:350) ~[factory.jar:master]
	at com.fasterxml.jackson.dataformat.xml.XmlFactory._createParser(XmlFactory.java:582) ~[factory.jar:master]
	at com.fasterxml.jackson.dataformat.xml.XmlFactory.createParser(XmlFactory.java:464) ~[factory.jar:master]
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3205) ~[factory.jar:master]
	at com.azure.core.util.serializer.JacksonAdapter.deserialize(JacksonAdapter.java:179) ~[factory.jar:master]
	at com.azure.core.implementation.serializer.HttpResponseBodyDecoder.deserializeBody(HttpResponseBodyDecoder.java:176) ~[factory.jar:master]
	at com.azure.core.implementation.serializer.HttpResponseBodyDecoder.lambda$decode$1(HttpResponseBodyDecoder.java:88) ~[factory.jar:master]
	... 50 more
```

**To Reproduce**
Simply use the `blobContainerClient.getBlobContainerurl()` or `blobServiceClient.generateAccountSas(..)` and it should happen with the response.

***Code Snippet***
```
var blobClient = new BlobServiceClientBuilder().connectionString("connectionStr").buildClient();
AccountSasResourceType resourceTypes = new AccountSasResourceType().setObject(true);
AccountSasPermission accountPermissions =
new AccountSasPermission().setReadPermission(true).setListPermission(true).setWritePermission(true);
AccountSasService accountService = new AccountSasService().setBlobAccess(true);
ZoneId tz = ZoneOffset.UTC.normalized();
OffsetDateTime expiration = OffsetDateTime.now(tz).plusHours(4);
blobClient.generateAccountSas(new AccountSasSignatureValues(expiration, accountPermissions, accountService, resourceTypes));
```
or

```
var blobClient = new BlobServiceClientBuilder().connectionString("connectionStr").buildClient();
var container = blobClient.getBlobContainerClient("containerName");
container.getBlobContainerUrl()
```

**Expected behavior**
The results should be returned from the API and the SDK as it did in v12.3.0 ..

**Setup (please complete the following information):**
 - OS: Linux
 - IDE : Intellij (but not relevant for this bug)
 - 12.5.0

**Additional context**
The exception shows a few different clues:
1. ```com.azure.core.implementation.serializer.MalformedValueException: Unexpected first character (char code 0xEF), not valid in xml document: could be mangled UTF-8 BOM marker. Make sure that the Reader uses correct encoding or pass an InputStream instead```

2. ```com.fasterxml.jackson.core.JsonParseException: Unexpected first character (char code 0xEF), not valid in xml document: could be mangled UTF-8 BOM marker. Make sure that the Reader uses correct encoding or pass an InputStream instead```

3. ```com.ctc.wstx.exc.WstxIOException: Unexpected first character (char code 0xEF), not valid in xml document: could be mangled UTF-8 BOM marker. Make sure that the Reader uses correct encoding or pass an InputStream instead ```

**Information Checklist**
Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [ x ] Bug Description Added
- [ x ] Repro Steps Added
- [ x ] Setup information Added

**Comment by alzimmermsft** [comment]
Hi @bryandmc, a newer version of Azure Storage Blobs was released this week, do you mind checking if this can be reproduced in the newer version as well?

https://search.maven.org/artifact/com.azure/azure-storage-blob/12.6.0/jar

**Comment by m-jarzebowski** [comment]
The newest version leads to the same issue. I investigated it a bit and it is related to the default Java encoding.
Here ([link](https://github.com/Azure/azure-sdk-for-java/blame/master/sdk/core/azure-core/src/main/java/com/azure/core/util/serializer/JacksonAdapter.java#L69)) you use _Charset.defaultCharset()_ to prepare _BOM_STRING_ to remove it from the response body. BOM_STRING does not match first characters of _value_ var ([link](https://github.com/Azure/azure-sdk-for-java/blame/master/sdk/core/azure-core/src/main/java/com/azure/core/util/serializer/JacksonAdapter.java#L172)) when java encoding is US-ASCII (probably for other encoding there is the same issue).
In the debugger i can see `value = "ï»¿<?xml version="1.0" encoding="utf-8"?><EnumerationResults ServiceEndp...`

I workaround it by setting Java default encoding (_-Dfile.encoding=UTF-8_). Then `value = "<?xml version="1.0" encoding="utf-8"?><EnumerationResults ServiceEndp...`

Side note:  _SerializerEncoding_ ([link](https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/core/azure-core/src/main/java/com/azure/core/util/serializer/SerializerEncoding.java#L12)) has nothing to do with encoding it is rather content type enum, it is quite confusing.

**Comment by alzimmermsft** [comment]
@m-jarzebowski thank you for the detailed investigation, that helped a lot in reproducing the issue. I have opened PR #10027 which should be a more robust fix to the issue.

**Comment by m-jarzebowski** [comment]
Thanks for the fix!

**Comment by alzimmermsft** [comment]
Closing as fixed by #10027

**Comment by nychi713** [comment]
I'm using sbt dependencies and with the azure-storage-blob , version 12.6.0 is still failing with the same error that @bryandmc  mentioned . I saw this fix was released recently into the azure-core sdk. However, I'm wondering if a new azure-storage-blob version needs to be released to catch that fix, due to in my case I use directly azure-storage-blob dependency. In the meantime, I applied the default encoding -Dfile.encoding=UTF-8 fix.
